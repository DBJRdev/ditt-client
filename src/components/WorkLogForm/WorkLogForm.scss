.formErrorText {
  text-align: center;
  color: #a32100;
}

.fieldset {
  max-width: 19em;
  padding: 0;
  margin-right: auto;
  margin-left: auto;
  border: 0;
}

.field {
  display: inline-block;
  min-width: 8.5rem;
}

.fieldDate {
  min-width: 3em;
}

.fieldDate input {
  width: 3em;
  text-align: center;
}

.formInfoText {
  font-size: 0.75rem;
  text-align: justify;
  color: #666;
}
