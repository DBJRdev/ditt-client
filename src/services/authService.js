import decode from 'jsonwebtoken/decode';
import {
  selectJwtToken,
  selectLastRequestBrowserTime,
} from '../resources/auth';
import { refreshToken } from '../resources/auth/actions';

export const registerAuthRefreshTokenService = (store) => {
  const authRefreshTokenService = () => {
    const {
      dispatch,
      getState,
    } = store;

    const lastRequestBrowserTime = selectLastRequestBrowserTime(getState());
    const jwtToken = selectJwtToken(getState());
    const jwtTokenPayload = decode(jwtToken);

    if (jwtTokenPayload === null) {
      return;
    }

    const {
      exp,
      iat,
    } = jwtTokenPayload;
    const now = Date.now() / 1000;

    if (exp - now <= 0) {
      return;
    }

    if (
      exp - now <= 300
      && lastRequestBrowserTime != null
      && ((lastRequestBrowserTime / 1000) - iat) > 15
    ) {
      dispatch(refreshToken());
    }
  };

  setInterval(() => authRefreshTokenService(), 15000);
};
