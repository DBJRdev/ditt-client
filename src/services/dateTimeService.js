import moment from 'moment-timezone';

const TIMEZONE = 'Europe/Prague';

export const localizedMoment = () => moment().tz(TIMEZONE);

export const isOverlapping = (dt1Start, dt1End, dt2Start, dt2End) => {
  if (
    dt1Start && moment.isMoment(dt1Start) && dt1Start.isValid()
    && dt1End && moment.isMoment(dt1End) && dt1End.isValid()
    && dt2Start && moment.isMoment(dt2Start) && dt2Start.isValid()
    && dt2End && moment.isMoment(dt2End) && dt2End.isValid()
  ) {
    return (dt1Start > dt2Start && dt1Start < dt2End)
      || (dt2Start > dt1Start && dt2Start < dt1End)
      || (dt1Start.diff(dt2Start, 'minutes') === 0 && dt1End.diff(dt2End, 'minutes') === 0);
  }

  throw new Error('Invalid datetime');
};

export const isWeekend = (momentDateTime) => {
  if (momentDateTime && moment.isMoment(momentDateTime) && momentDateTime.isValid()) {
    return momentDateTime.day() === 0 || momentDateTime.day() === 6;
  }

  throw new Error('Invalid datetime');
};

export const toJson = momentDateTime => momentDateTime.local().format();

export const toMomentDateTime = (dateTimeString) => {
  const momentDateTime = moment(dateTimeString, moment.ISO_8601).tz(TIMEZONE);

  if (momentDateTime.isValid()) {
    return momentDateTime;
  }

  throw new Error('Invalid datetime');
};

export const toDayFormat = (momentDateTime) => {
  if (momentDateTime && moment.isMoment(momentDateTime) && momentDateTime.isValid()) {
    return momentDateTime.format('dddd');
  }

  throw new Error('Invalid datetime');
};

export const toHourMinuteFormat = (momentDateTime) => {
  if (momentDateTime && moment.isMoment(momentDateTime) && momentDateTime.isValid()) {
    return momentDateTime.format('HH:mm');
  }

  throw new Error('Invalid datetime');
};

export const toMonthYearFormat = (momentDateTime) => {
  if (momentDateTime && moment.isMoment(momentDateTime) && momentDateTime.isValid()) {
    return momentDateTime.format('MMMM YYYY');
  }

  throw new Error('Invalid datetime');
};

export const toDayMonthYearFormat = (momentDateTime) => {
  if (momentDateTime && moment.isMoment(momentDateTime) && momentDateTime.isValid()) {
    return momentDateTime.format('L');
  }

  throw new Error('Invalid datetime');
};
