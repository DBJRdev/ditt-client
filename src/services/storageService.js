const workLogTimerKey = 'workLogTimer';

export const getWorkLogTimer = () => localStorage.getItem(workLogTimerKey);

export const removeWorkLogTimer = () => {
  localStorage.removeItem(workLogTimerKey);
};

export const setWorkLogTimer = (value) => {
  localStorage.setItem(workLogTimerKey, value);
};
