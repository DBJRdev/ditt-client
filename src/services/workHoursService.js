export const getWorkHoursString = (workHoursValue) => {
  const hour = Math.floor(workHoursValue);
  let minute = 0;

  if (hour > 0) {
    minute = Math.round((workHoursValue % hour) * 60);
  } else {
    minute = Math.round(workHoursValue * 60);
  }

  if (minute === 0) {
    return `${hour}:00`;
  } else if (minute < 10) {
    return `${hour}:0${minute}`;
  }

  return `${hour}:${minute}`;
};

export const getWorkHoursValue = (workHoursString) => {
  const splittedString = workHoursString.split(':');

  const hour = parseInt(splittedString[0], 10);
  const minute = parseInt(splittedString[1], 10);

  return hour + (minute / 60);
};
