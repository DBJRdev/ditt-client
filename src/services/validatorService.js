import validator from 'validator';
import { isOverlapping } from './dateTimeService';

const validateWorkLog = (workLog, workLogsOfDay) => {
  const errors = {
    elements: {
      endHour: null,
      endMinute: null,
      form: null,
      startHour: null,
      startMinute: null,
    },
    isValid: true,
  };

  const hourCheck = [
    'endHour',
    'startHour',
  ];

  const minuteCheck = [
    'endMinute',
    'startMinute',
  ];

  [...hourCheck, ...minuteCheck].forEach((element) => {
    if (
      workLog[element] === null
      || validator.isEmpty(workLog[element].toString())
      || !validator.isNumeric(workLog[element].toString())
    ) {
      errors.elements[element] = 'Not a number.';
      errors.isValid = false;
    }
  });

  if (!errors.isValid) {
    return errors;
  }

  hourCheck.forEach((element) => {
    if (workLog[element] < 0 || workLog[element] > 23) {
      errors.elements[element] = 'Invalid hour.';
      errors.isValid = false;
    }
  });

  minuteCheck.forEach((element) => {
    if (workLog[element] < 0 || workLog[element] > 59) {
      errors.elements[element] = 'Invalid minute.';
      errors.isValid = false;
    }
  });

  if (!errors.isValid) {
    return errors;
  }

  if (
    (workLog.startHour * 60) + workLog.startMinute >= (workLog.endHour * 60) + workLog.endMinute
  ) {
    errors.elements.endHour = 'Invalid hour';
    errors.elements.endMinute = 'Invalid minute';
    errors.isValid = false;
  }

  if (!errors.isValid) {
    return errors;
  }

  if (workLogsOfDay.length > 0) {
    let overlapping = false;
    const todayDate = workLogsOfDay[0].startTime;
    const startTime = todayDate.clone().hour(workLog.startHour).minute(workLog.startMinute);
    const endTime = todayDate.clone().hour(workLog.endHour).minute(workLog.endMinute);

    workLogsOfDay.forEach((workLogItem) => {
      if (isOverlapping(startTime, endTime, workLogItem.startTime, workLogItem.endTime)) {
        overlapping = true;
      }
    });

    if (overlapping) {
      errors.elements.form = 'Entered worklog overlaps existing one.';
      errors.isValid = false;
    }
  }

  return errors;
};

export default validateWorkLog;
