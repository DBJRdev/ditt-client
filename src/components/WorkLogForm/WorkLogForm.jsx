import moment from 'moment';
import PropTypes from 'prop-types';
import React from 'react';
import ImmutablePropTypes from 'react-immutable-proptypes';
import {
  Modal,
  TextField,
} from 'react-ui';
import {
  ADD_WORK_LOG_FAILURE,
  ADD_WORK_LOG_SUCCESS,
} from '../../resources/workLog/actionTypes';
import { toDayMonthYearFormat } from '../../services/dateTimeService';
import { validateWorkLog } from '../../services/validatorService';

class WorkLogForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      formData: {
        endHour: props.date.hour(),
        endMinute: props.date.minute(),
        startHour: props.date.hour(),
        startMinute: props.date.minute(),
      },
      formValidity: {
        elements: {
          endHour: null,
          endMinute: null,
          form: null,
          startHour: null,
          startMinute: null,
        },
        isValid: false,
      },
    };

    this.changeHandler = this.changeHandler.bind(this);
    this.saveHandler = this.saveHandler.bind(this);

    this.formErrorStyle = {
      color: '#a32100',
      textAlign: 'center',
    };

    this.fieldSetStyle = {
      borderStyle: 'none',
      marginLeft: 'auto',
      marginRight: 'auto',
      maxWidth: '17em',
      padding: '0',
    };

    this.fieldStyle = {
      display: 'inline-block',
    };
  }

  changeHandler(e) {
    const eventTarget = e.target;

    this.setState((prevState) => {
      const formData = Object.assign({}, prevState.formData);
      formData[eventTarget.id] = eventTarget.value;

      return { formData };
    });
  }

  saveHandler() {
    const { date } = this.props;
    const { formData } = this.state;
    const formValidity = validateWorkLog(formData, this.props.workLogsOfDay.toJS());

    this.setState({ formValidity });

    if (formValidity.isValid) {
      this.props.saveHandler({
        endTime: date.clone().hour(formData.endHour).minute(formData.endMinute).second(0),
        startTime: date.clone().hour(formData.startHour).minute(formData.startMinute).second(0),
      })
        .then((response) => {
          if (response.type === ADD_WORK_LOG_SUCCESS) {
            this.props.closeHandler();
          } else if (response.type === ADD_WORK_LOG_FAILURE) {
            formValidity.elements.form = 'Work log cannot be added.';

            this.setState({ formValidity });
          }
        });
    }
  }

  render() {
    return (
      <Modal
        actions={[
          {
            clickHandler: this.saveHandler,
            label: 'Save',
            loading: this.props.isPosting,
          },
        ]}
        closeHandler={this.props.closeHandler}
        title="Add work log"
      >
        <p style={this.formErrorStyle}>
          {this.state.formValidity.elements.form}
        </p>
        <form>
          <fieldset style={this.fieldSetStyle}>
            <legend>Date</legend>
            <div style={this.fieldStyle}>
              <TextField
                disabled
                fieldId="date"
                isLabelVisible={false}
                label="Date"
                value={toDayMonthYearFormat(this.props.date) || ''}
              />
            </div>
          </fieldset>
          <fieldset style={this.fieldSetStyle}>
            <legend>Start time:</legend>
            <div>
              <div style={this.fieldStyle}>
                <TextField
                  autoFocus
                  changeHandler={this.changeHandler}
                  error={this.state.formValidity.elements.startHour}
                  fieldId="startHour"
                  isLabelVisible={false}
                  label="Start hour"
                  value={this.state.formData.startHour || ''}
                />
              </div>
              :
              <div style={this.fieldStyle}>
                <TextField
                  changeHandler={this.changeHandler}
                  error={this.state.formValidity.elements.startMinute}
                  fieldId="startMinute"
                  isLabelVisible={false}
                  label="Start minute"
                  value={this.state.formData.startMinute || ''}
                />
              </div>
              &nbsp;h
            </div>
          </fieldset>
          <fieldset style={this.fieldSetStyle}>
            <legend>End time:</legend>
            <div>
              <div style={this.fieldStyle}>
                <TextField
                  changeHandler={this.changeHandler}
                  error={this.state.formValidity.elements.endHour}
                  fieldId="endHour"
                  isLabelVisible={false}
                  label="End hour"
                  value={this.state.formData.endHour || ''}
                />
              </div>
              :
              <div style={this.fieldStyle}>
                <TextField
                  changeHandler={this.changeHandler}
                  error={this.state.formValidity.elements.endMinute}
                  fieldId="endMinute"
                  isLabelVisible={false}
                  label="End minute"
                  value={this.state.formData.endMinute || ''}
                />
              </div>
              &nbsp;h
            </div>
          </fieldset>
        </form>
      </Modal>
    );
  }
}

WorkLogForm.propTypes = {
  closeHandler: PropTypes.func.isRequired,
  date: PropTypes.instanceOf(moment).isRequired,
  isPosting: PropTypes.bool.isRequired,
  saveHandler: PropTypes.func.isRequired,
  workLogsOfDay: ImmutablePropTypes.listOf(ImmutablePropTypes.mapContains({
    endTime: PropTypes.shape.isRequired,
    id: PropTypes.number.isRequired,
    startTime: PropTypes.shape.isRequired,
  })).isRequired,
};

export default WorkLogForm;
