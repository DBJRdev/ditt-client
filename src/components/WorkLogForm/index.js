import WorkLogForm from './WorkLogForm';

export default WorkLogForm;
