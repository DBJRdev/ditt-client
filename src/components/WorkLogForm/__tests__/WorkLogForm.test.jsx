import Immutable from 'immutable';
import React from 'react';
import sinon from 'sinon';
import { mount } from 'enzyme';
import WorkLogForm from '../WorkLogForm';
import { localizedMoment } from '../../../services/dateTimeService';

let fakeTimestamp = null;
let fakeMomentDateTime = null;

beforeEach(() => {
  fakeTimestamp = sinon.useFakeTimers(1514764800000);
  fakeMomentDateTime = localizedMoment();
});

afterEach(() => {
  fakeTimestamp.restore();
});

describe('rendering', () => {
  it('renders correctly with mandatory props', () => {
    const tree = mount(<WorkLogForm
      banWorkLogsOfDay={Immutable.List([])}
      config={Immutable.Map()}
      date={fakeMomentDateTime}
      isPosting={false}
      onClose={() => {}}
      onSave={() => {}}
      user={Immutable.fromJS({
        vacations: [
          {
            remainingVacationDays: 20,
            year: 2017,
          },
          {
            remainingVacationDays: 20,
            year: 2018,
          },
        ],
      })}
      workLogsOfDay={Immutable.List([])}
    />);

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = mount(<WorkLogForm
      banWorkLogsOfDay={Immutable.List([])}
      config={Immutable.Map()}
      date={fakeMomentDateTime}
      isPosting={false}
      onClose={() => {}}
      onSave={() => {}}
      showInfoText
      user={Immutable.fromJS({
        vacations: [
          {
            remainingVacationDays: 20,
            year: 2017,
          },
          {
            remainingVacationDays: 20,
            year: 2018,
          },
        ],
      })}
      workLogsOfDay={Immutable.List([])}
    />);

    expect(tree).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls onClose() date when previous month button is clicked', () => {
    const spy = sinon.spy();
    const tree = mount(<WorkLogForm
      banWorkLogsOfDay={Immutable.List([])}
      config={Immutable.Map()}
      date={fakeMomentDateTime}
      isPosting={false}
      onClose={spy}
      onSave={() => {}}
      user={Immutable.fromJS({
        vacations: [
          {
            remainingVacationDays: 20,
            year: 2017,
          },
          {
            remainingVacationDays: 20,
            year: 2018,
          },
        ],
      })}
      workLogsOfDay={Immutable.List([])}
    />);

    tree.find('button').last().simulate('click');

    expect(spy.calledOnce).toEqual(true);
  });
});
