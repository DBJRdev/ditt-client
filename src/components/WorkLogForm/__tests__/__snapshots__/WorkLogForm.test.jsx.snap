// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly with all props 1`] = `
<Modal
  actions={
    Array [
      Object {
        "clickHandler": [Function],
        "label": "Save",
        "loading": false,
      },
    ]
  }
  closeHandler={[Function]}
  title="Add work log"
>
  <p
    style={
      Object {
        "color": "#a32100",
        "textAlign": "center",
      }
    }
  />
  <form>
    <fieldset
      style={
        Object {
          "borderStyle": "none",
          "marginLeft": "auto",
          "marginRight": "auto",
          "maxWidth": "17em",
          "padding": "0",
        }
      }
    >
      <legend>
        Date
      </legend>
      <div
        style={
          Object {
            "display": "inline-block",
          }
        }
      >
        <TextField
          changeHandler={null}
          description={null}
          disabled={true}
          error={null}
          fieldId="date"
          isLabelVisible={false}
          label="Date"
          placeholder={null}
          required={false}
          type="text"
          value="01/01/2018"
        />
      </div>
    </fieldset>
    <fieldset
      style={
        Object {
          "borderStyle": "none",
          "marginLeft": "auto",
          "marginRight": "auto",
          "maxWidth": "17em",
          "padding": "0",
        }
      }
    >
      <legend>
        Start time:
      </legend>
      <div>
        <div
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <TextField
            autoFocus={true}
            changeHandler={[Function]}
            description={null}
            disabled={false}
            error={null}
            fieldId="startHour"
            isLabelVisible={false}
            label="Start hour"
            placeholder={null}
            required={false}
            type="text"
            value={1}
          />
        </div>
        :
        <div
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <TextField
            changeHandler={[Function]}
            description={null}
            disabled={false}
            error={null}
            fieldId="startMinute"
            isLabelVisible={false}
            label="Start minute"
            placeholder={null}
            required={false}
            type="text"
            value=""
          />
        </div>
         h
      </div>
    </fieldset>
    <fieldset
      style={
        Object {
          "borderStyle": "none",
          "marginLeft": "auto",
          "marginRight": "auto",
          "maxWidth": "17em",
          "padding": "0",
        }
      }
    >
      <legend>
        End time:
      </legend>
      <div>
        <div
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <TextField
            changeHandler={[Function]}
            description={null}
            disabled={false}
            error={null}
            fieldId="endHour"
            isLabelVisible={false}
            label="End hour"
            placeholder={null}
            required={false}
            type="text"
            value={1}
          />
        </div>
        :
        <div
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <TextField
            changeHandler={[Function]}
            description={null}
            disabled={false}
            error={null}
            fieldId="endMinute"
            isLabelVisible={false}
            label="End minute"
            placeholder={null}
            required={false}
            type="text"
            value=""
          />
        </div>
         h
      </div>
    </fieldset>
  </form>
</Modal>
`;
